src/audio/drc/drc_algorithm_hifi3.c

```diff
@@ -0,0 +1,469 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2020 Google LLC. All rights reserved.
+//
+// Author: Pin-chih Lin <johnylin@google.com>
+
+#include <sof/audio/component.h>
+#include <sof/audio/drc/drc.h>
+#include <sof/audio/drc/drc_algorithm.h>
+#include <sof/audio/drc/drc_math.h>
+#include <sof/audio/format.h>
+#include <sof/math/decibels.h>
+#include <sof/math/numbers.h>
+#include <stdint.h>
+
+#if DRC_HIFI3
+
+#include <xtensa/tie/xt_hifi3.h>
+
+#define ONE_Q20        1048576    /* Q_CONVERT_FLOAT(1.0f, 20) */
+#define ONE_Q21        2097152    /* Q_CONVERT_FLOAT(1.0f, 21) */
+#define ONE_Q30        1073741824 /* Q_CONVERT_FLOAT(1.0f, 30) */
+#define TWELVE_Q21     25165824   /* Q_CONVERT_FLOAT(12.0f, 21) */
+#define HALF_Q24       8388608    /* Q_CONVERT_FLOAT(0.5f, 24) */
+#define NEG_TWO_DB_Q30 852903424  /* Q_CONVERT_FLOAT(0.7943282347242815f, 30) */
+
+/*
+ * A substitutive function of Q_MULTSR_32X32(a, b, qa, qb, qy) in HIFI manners.
+ *
+ * In AE_MULF32R, it takes the multiplication of 1.31 and 1.31 to 17.47, which means it does
+ * right-shift 15 bits. Let's consider a is .qa, b is .qb, then tmp will be .(qa+qb-15)
+ * In AE_ROUND32F48SSYM, it rounds 17.47 to 1.31, equally right-shifts 16 bits. Let's consider the
+ * output is .qy, then tmp here needs to be .(qy+16)
+ * If we set the left-shift bit from the former tmp to the latter tmp as "x", then:
+ *     (qa+qb-15) + x = (qy+16)
+ *     x = qy-qa-qb+31
+ */
+static inline ae_f32 mult(ae_f32 a, ae_f32 b, int32_t qa, int32_t qb, int32_t qy)
```

Comment:
we should evaluate what math functions like this one are generic enough to move to lib/math for reuse